Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -921,7 +921,7 @@ - (GCFSStructuredQuery_Filter *)encodedUnaryFilter:(id<FSTFilter>)filter {
   } else if ([filter isKindOfClass:[FSTNullFilter class]]) {
     proto.unaryFilter.op = GCFSStructuredQuery_UnaryFilter_Operator_IsNull;
   } else {
-    FSTFail(@"Unrecognized filter: %@", filter);
+    HARD_FAIL("Unrecognized filter: %s", static_cast<id>(filter));
```

Comment:
Why is this `static_cast` necessary?

Reply:
The static cast is necessary because the type `id` is magic in a way that Objective-C++ only halfway understands.

`id` in Objective-C is a fully dynamic pointer. It's handled specially by the compiler and all method invocations against it are resolved at run-time only.

Objective-C also has the ability to annotate pointers of type `id` to indicate that the pointed to object conforms to a list of protocols. It looks like `id<FSTFilter>`.

However, `id` is declared as follows:

```c
typedef struct objc_object *id;
```

In Objective-C++ the dynamic resolution behavior of `id` is preserved, but `id` and `id<FSTFilter>` are treated as distinct for overload resolution purposes. This creates an ambiguity:

```c++
  FormatArg(id object, internal::FormatChoice<0>)
```
vs
```c++
  template <typename T>
  FormatArg(T&& value, internal::FormatChoice<4>)
```
The former does not exactly match `id<FSTFilter>`, but matches `FormatChoice<0>{}` exactly. The latter matches `id<FSTFilter>` exactly but does not exactly match `FormatChoice<0>{}`.

Unfortunately you can't write something like this:
```c++
  template <typename T>
  FormatArg(id<T> object, internal::FormatChoice<0>)
```
because `id` is not a generic class and the angle brackets aren't template parameters. `id` cannot be operated on in a generic way outside an Objective-C `@interface` block. This means that the traditional C++ way of forcing the resolution of the ambiguity by making one of the arguments a template does not work.

This does not work either:
```c++
  template <
      typename T,
      typename = typename std::enable_if<std::is_base_of<id, T>{}>::type>
  FormatArg(id object, internal::FormatChoice<0>)
```
since `id` is not a class type as far as C++ is concerned.

It seems that while the compiler recognizes that `id` and `id<FSTFilter>` are related types and that something of type `id<FSTFilter>` would be suitable to pass to something that accepts `id`, I can't find a way to express a preference in the C++ type system for one FormatArg constructor over the other.

Since we don't use this frequently I figured a static cast in the caller is a suitable workaround. If you have any other ideas about how to handle this I'm open to them.

Reply:
Thanks for the explanation! I'm totally fine with it. Even assuming there is a better way of dealing with this (which isn't necessarily true), I don't think it's worthwhile trying to figure it out because this interoperability should become unnecessary in the future.